scheds/rust/scx_rusty/src/bpf/main.bpf.c

```diff
@@ -547,6 +548,57 @@ static u64 sched_prio_to_latency_weight(u64 prio)
 	return sched_prio_to_weight[DL_MAX_LAT_PRIO - prio - 1];
 }
 
+
+/*
+ * Returns the preferred domain according to the mempolicy. See man(2)
+ * set_mempolicy for more details.
+ */
+static u32 task_pick_mempolicy_domain(
+		struct task_struct *p, struct task_ctx *taskc, u32 rr_token)
+{
+	u32 ret = NO_DOM_FOUND;
+	if (!mempolicy_affinity || !bpf_core_field_exists(p->mempolicy) ||
+			!p->mempolicy || !taskc->cpumask)
+		return ret;
+
+	if (!(p->mempolicy->mode & (MPOL_BIND|MPOL_PREFERRED))) {
+		return ret;
+	}
+
+	// MPOL_BIND and MPOL_PREFERRED_MANY use the home_node field on the
+	// mempolicy struct, so use that for now. In the future the memory
+	// usage of the node can be checked to follow the same algorithm for
+	// where memory allocations will occur.
+	if ((int)p->mempolicy->home_node >= 0) {
+		struct node_ctx *nodec;
+		return (u32)p->mempolicy->home_node;
+	}
+
+	nodemask_t *node_mask = &p->mempolicy->nodes;
+
+	void *mask = BPF_CORE_READ(node_mask, bits);
+	u32 val = 0;
+	if (bpf_core_read(&val, sizeof(val), mask)) {
+		return ret;
+	}
+
+	bool cleared = false;
+	u32 i;
+	bpf_for(i, 0, nr_nodes) {
+		if (!(val & (1 <<i))) {
+			continue;
+		}
+
+		ret = i;
+		// If the round robin token matches the domain then return
+		// it.
+		if (ret != NO_DOM_FOUND && i % rr_token == 0)
+			return ret;
```

Comment:
Spoke offline -- we're matching NUMA node with domain ID here, but they might not match up if there are multiple CCXs per node. We can instead have something like a mask of preferred domains (that's informed by `p->mempolicy->nodes`), and then query that when looking for a domain when we do round robin in the caller, and in the load balancer.